polaris-core/src/main/java/org/apache/polaris/core/persistence/pagination/Page.java

```diff
@@ -18,25 +18,96 @@
+  @Nullable private final String encodedDataReference;
 
-  public Page(PageToken pageToken, List<T> items) {
-    this.pageToken = pageToken;
+  private Page(PageToken request, @Nullable String encodedDataReference, List<T> items) {
+    this.request = request;
+    this.encodedDataReference = encodedDataReference;
     this.items = items;
   }
 
   /**
-   * Used to wrap a {@link List<T>} of items into a {@link Page <T>} when there are no more pages
+   * Builds a complete response page for the full list of relevant items. No subsequence pages of
+   * related data exist.
    */
   public static <T> Page<T> fromItems(List<T> items) {
-    return new Page<>(new DonePageToken(), items);
+    return new Page<>(PageToken.readEverything(), null, items);
+  }
+
+  /**
+   * Produces a response page by consuming the number of items from the provided stream according to
+   * the {@code request} parameter. Source items can be converted to a different type by providing a
+   * {@code mapper} function. The page token for the response will be produced from the request data
+   * combined with the pointer to the next page of data provided by the {@code dataPointer}
+   * function.
+   *
+   * @param request defines pagination parameters that were uses to produce this page of data.
+   * @param items stream of source data
+   * @param mapper converter from source data types to response data types.
+   * @param dataPointer determines the internal pointer to the next page of data given the last item
+   *     from the previous page. The output of this function will be available from {@link
+   *     PageToken#encodedDataReference()} associated with the request for the next page.
+   */
+  public static <R, T> Page<R> mapped(
+      PageToken request, Stream<T> items, Function<T, R> mapper, Function<T, String> dataPointer) {
+    List<R> data;
+    if (request.paginationRequested()) {
+      data = new ArrayList<>(request.pageSize());
+    } else {
+      data = new ArrayList<>();
+    }
+
+    T last = null;
+    Iterator<T> it = iterator(items.spliterator());
+    while (it.hasNext() && (!request.paginationRequested() || data.size() < request.pageSize())) {
+      last = it.next();
+      data.add(mapper.apply(last));
+    }
+
+    if (request.paginationRequested() && data.size() < request.pageSize()) {
+      // the page was not filled, inform the client that there's no next page
+      last = null;
+    }
+
+    return new Page<>(request, dataPointer.apply(last), data);
+  }
+
+  public List<T> items() {
+    return items;
+  }
+
+  /**
+   * Returns a page token in encoded form suitable for returning to API clients. The string returned
+   * from this method is expected to be parsed by {@link PageToken#decode(String, Integer)} when
+   * servicing the request for the next page of related data.
+   */
+  public @Nullable String encodedResponseToken() {
+    return PageTokenUtil.encodePageToken(request, encodedDataReference);
```

Comment:
Also, this looks like quite a thin method

Reply:
Yes, it does :)

Reply:
I can make this change if you prefer. No worries.